xdr/next/Stellar-contract.x

```diff
@@ -71,16 +71,114 @@ enum SCStatic
 enum SCStatusType
 {
     SST_OK = 0,
-    SST_UNKNOWN_ERROR = 1
+    SST_UNKNOWN_ERROR = 1,
+    SST_HOST_VALUE_ERROR = 2,
+    SST_HOST_OBJECT_ERROR = 3,
+    SST_HOST_FUNCTION_ERROR = 4,
+    SST_HOST_STORAGE_ERROR = 5,
+    SST_HOST_CONTEXT_ERROR = 6,
+    SST_VM_ERROR = 7
     // TODO: add more
 };
 
+enum SCHostValErrorCode
+{
+    HOST_VALUE_UNKNOWN_ERROR = 0,
+    HOST_VALUE_RESERVED_TAG_VALUE = 1,
+    HOST_VALUE_UNEXPECTED_VAL_TYPE = 2,
+    HOST_VALUE_U63_OUT_OF_RANGE = 3,
+    HOST_VALUE_U32_OUT_OF_RANGE = 4,
+    HOST_VALUE_STATIC_UNKNOWN = 5,
+    HOST_VALUE_MISSING_OBJECT = 6,
+    HOST_VALUE_SYMBOL_TOO_LONG = 7,
+    HOST_VALUE_SYMBOL_BAD_CHAR = 8,
+    HOST_VALUE_SYMBOL_CONTAINS_NON_UTF8 = 9,
+    HOST_VALUE_BITSET_TOO_MANY_BITS = 10,
+    HOST_VALUE_STATUS_UNKNOWN = 11
+};
+
+enum SCHostObjErrorCode
+{
+    HOST_OBJECT_UNKNOWN_ERROR = 0,
+    HOST_OBJECT_UNKNOWN_REFERENCE = 1,
+    HOST_OBJECT_UNEXPECTED_TYPE = 2,
+    HOST_OBJECT_OBJECT_COUNT_EXCEEDS_U32_MAX = 3,
+    HOST_OBJECT_OBJECT_NOT_EXIST = 4,
+    HOST_OBJECT_VEC_INDEX_OUT_OF_BOUND = 5,
+    HOST_OBJECT_VEC_ELEMENT_NOT_EXIST = 6,
+    HOST_OBJECT_CONTRACT_HASH_WRONG_LENGTH = 7
+};
+
+enum SCHostFnErrorCode
+{
+    HOST_FN_UNKNOWN_ERROR = 0,
+    HOST_FN_UNEXPECTED_HOST_FUNCTION_ACTION = 1,
+    HOST_FN_INPUT_ARGS_WRONG_LENGTH = 2,
+    HOST_FN_INPUT_ARGS_WRONG_TYPE = 3,
+    HOST_FN_INPUT_ARGS_INVALID = 4
+};
+
+enum SCHostStorageErrorCode
+{
+    HOST_STORAGE_UNKNOWN_ERROR = 0,
+    HOST_STORAGE_EXPECT_CONTRACT_DATA = 1,
+    HOST_STORAGE_READWRITE_ACCESS_TO_READONLY_ENTRY = 2,
+    HOST_STORAGE_ACCESS_TO_UNKNOWN_ENTRY = 3,
+    HOST_STORAGE_MISSING_KEY_IN_GET = 4,
+    HOST_STORAGE_GET_ON_DELETED_KEY = 5
+};
+
+enum SCHostContextErrorCode
+{
+    HOST_CONTEXT_UNKNOWN_ERROR = 0,
+    HOST_CONTEXT_NO_CONTRACT_RUNNING = 1
+};
+
+enum SCVmErrorCode {
+    WASMI_ERROR_UNKNOWN = 0,
+    WASMI_ERROR_VALIDATION = 1,
+    WASMI_ERROR_INSTANTIATION = 2,
+    WASMI_ERROR_FUNCTION = 3,
+    WASMI_ERROR_TABLE = 4,
+    WASMI_ERROR_MEMORY = 5,
+    WASMI_ERROR_GLOBAL = 6,
+    WASMI_ERROR_VALUE = 7,
+    WASMI_TRAP_UNREACHABLE = 8,
+    WASMI_TRAP_MEMORY_ACCESS_OUT_OF_BOUNDS = 9,
+    WASMI_TRAP_TABLE_ACCESS_OUT_OF_BOUNDS = 10,
+    WASMI_TRAP_ELEM_UNINITIALIZED = 11,
+    WASMI_TRAP_DIVISION_BY_ZERO = 12,
+    WASMI_TRAP_INTEGER_OVERFLOW = 13,
+    WASMI_TRAP_INVALID_CONVERSION_TO_INT = 14,
+    WASMI_TRAP_STACK_OVERFLOW = 15,
+    WASMI_TRAP_UNEXPECTED_SIGNATURE = 16
```

Comment:
Could we normalize the prefix on these? Because some are WASMI_ERROR_ and others are WASMI_TRAP_ in generated code we end up with lots of ERROR_ and TRAP_.

Reply:
The reason I maintain these prefix is because this is how it's structured in WASMI, where there the `WASMI::Error::*` contains the `WASMI_ERROR_*` matching types, plus a `WASMI::Error::Trap::*` with sub-types for each trap. I'll rename the `WASMI_TRAP_` to `WASMI_ERROR_TRAP_` so that the `Error` prefix will not show up in the generated type.

Reply:
Interesting. If trap errors are a type of error, maybe we should treat them the same in the protocol and make the type of trap a subfield? I don't feel strongly but just pointing out we could do that.